python/cmsis_svd/model.py

```diff
@@ -1,452 +1,620 @@
-# You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#
-import json
-import six
+"""
+
+Copyright 2015-2024 cmsis-svd Authors
 
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
 
-def _check_type(value, expected_type):
-    """Perform type checking on the provided value
+    http://www.apache.org/licenses/LICENSE-2.0
 
-    This is a helper that will raise ``TypeError`` if the provided value is
-    not an instance of the provided type.  This method should be used sparingly
-    but can be good for preventing problems earlier when you want to restrict
-    duck typing to make the types of fields more obvious.
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
 
-    If the value passed the type check it will be returned from the call.
-    """
-    if not isinstance(value, expected_type):
-        raise TypeError("Value {value!r} has unexpected type {actual_type!r}, expected {expected_type!r}".format(
-            value=value,
-            expected_type=expected_type,
-            actual_type=type(value),
-        ))
-    return value
+"""
+
+from typing import List, Optional, Any, Union, TypedDict, Generator
+from dataclasses import dataclass, field, fields
+from enum import Enum
+import inspect
+import copy
+import json
+import six
 
 
-def _none_as_empty(v):
+################################################################################
+# Utils
+################################################################################
+
+def _none_as_empty(v: Any) -> Generator[Any, None, None]:
     if v is not None:
         for e in v:
             yield e
 
 
+def isinstance_by_str(instance: Any, type_name: str) -> bool:
+    if (cls := globals().get(type_name, None)) and inspect.isclass(cls):
+        return isinstance(instance, cls)
+    return False
+
+
+################################################################################
+# CMSIS-SVD Model Base class
+################################################################################
+
 class SVDJSONEncoder(json.JSONEncoder):
-    _TO_DICT_SKIP_KEYS = {"register_arrays", "parent"}
+    _TO_DICT_SKIP_KEYS = {'parent'}
 
-    def default(self, obj):
-        if isinstance(obj, SVDElement):
+    def default(self, o: Any) -> Any:
```

Comment:
nit: Why introduce this variable rename?

Reply:
Fixed